### PR TITLE
autoupgrade: use heartbeats to detect crashed autoupgrade attempts

### PR DESCRIPTION
**Pull request by Strum355**
If a frontend crashes due to OOM or any other reason, in a way that it fails to set the `success` column in `upgrade_logs` to `false`, then all frontends will block trying to claim the autoupgrade lock until the status is manually set in postgres (or the table is truncated). With a heartbeat mechanism, we can allow claiming the lock if the last lock hasnt received a heartbeat in a certain amount of time (here we're giving 6 missed heartbeats space).

## Test plan

Added unit tests for the claim query, tested in docker-compose
image

### CODE REVIEW

File: internal/version/upgradestore/store.go [hunk 1]
```diff
@@ -19,18 +20,20 @@ import (
...
-	db *basestore.Store
+	db    *basestore.Store
+	clock glock.Clock
 }
 
 // New returns a new version store with the given database handle.
-func New(db database.DB) *store {
-	return NewWith(db.Handle())
+func New(db database.DB, clock glock.Clock) *store {
+	return NewWith(db.Handle(), clock)
```
Comment (efritz) [inline internal/version/upgradestore/store.go:29]: If this clock is used only for testing, I'd suggest having a public `New` that calls a private `newStore` that always passes a real clock. That way external consumers don't have to create a real clock but the internal test package can use this alternative constructor that allows it to be mocked.

File: cmd/frontend/internal/cli/autoupgrade.go [hunk 2]
```diff
@@ -299,85 +293,37 @@ func claimAutoUpgradeLock(ctx context.Context, obsvCtx *observation.Context, db
...
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			case <-ticker.C:
+				func() {
+					ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+					defer cancel()
+					if err := upgradestore.Heartbeat(ctx); err != nil {
```
Comment (efritz) [inline cmd/frontend/internal/cli/autoupgrade.go:318]: Don't we only want to do this if we have an active claim?

Comment (efritz) [reply cmd/frontend/internal/cli/autoupgrade.go:318]: Awesome, forgot how the flow went and the diff is very collapsed by default.

**Comment by Strum355** [comment]
> LGTM, but make sure you test locally by killing an in-progress upgrade then showing that it's re-entrant.

Added screenie to original description

**Pull request merged**